### PR TITLE
Speed up greatCircleDistance function with test

### PR DESCRIPTION
**Pull request by stavrolia**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.
Speedup greatCircleDistance function with the code from Andrey Aksenov (for issue #5739)

Category (leave one):
    Performance Improvement

Short description (up to few sentences):
Speedup and test correctness of current function. Add perf test.

This PR made instead of #7254

### CODE REVIEW

**Comment by stavrolia** [comment]
Functional test for greatCircleDistance now look only on distance ~340 km. It is sad, because on such distance using haversine formula (old version) we don't know what exact amount of meters it is. I checked the pair of coordinates from test in more precise and slow formula - vincenty, and the distance was 344.480489697 km. With haversine formula (in test) it is 343417 m, with formula of this PR it is 344579 m. So new version is better in this case, and current test is awfull.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 1]
```diff
@@ -38,7 +119,7 @@ class FunctionGreatCircleDistance : public IFunction
 public:
 
     static constexpr auto name = "greatCircleDistance";
-    static FunctionPtr create(const Context &) { return std::make_shared<FunctionGreatCircleDistance>(); }
+    static FunctionPtr create(const Context &) { GeodistInit(); return std::make_shared<FunctionGreatCircleDistance>(); }
```
Comment (akuzm) [inline dbms/src/Functions/greatCircleDistance.cpp]: `create()` is called in a query context, so calling `GeodistInit()` here might introduce a ThreadSanitizer warning, if two queries with this functions are executed simultaneously. A good place to do the initialization of global constants would be `registerFunction`, which is called at server startup in a single-threaded context.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 2]
```diff
@@ -38,7 +119,7 @@ class FunctionGreatCircleDistance : public IFunction
 public:
 
     static constexpr auto name = "greatCircleDistance";
-    static FunctionPtr create(const Context &) { return std::make_shared<FunctionGreatCircleDistance>(); }
+    static FunctionPtr create(const Context &) {  return std::make_shared<FunctionGreatCircleDistance>(); }
```
Comment (akuzm) [inline dbms/src/Functions/greatCircleDistance.cpp]:
```suggestion
    static FunctionPtr create(const Context &) { return std::make_shared<FunctionGreatCircleDistance>(); }
```

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 3]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+        double x = PI * i / GEODIST_TABLE_K - PI * 0.5; // [-pi/2, pi/2] -> [0, KTABLE]
+        g_GeoFlatK[i][0] = static_cast<float>(sqr(111132.09 - 566.05 * cos(2 * x) + 1.20 * cos(4 * x)));
+        g_GeoFlatK[i][1] = static_cast<float>(sqr(111415.13 * cos(x) - 94.55 * cos(3 * x) + 0.12 * cos(5 * x)));
+    }
+}
+
+static inline float GeodistDegDiff(float f)
+{
+    f = static_cast<float>(fabs(f));
+    while (f > 360)
```
Comment (akuzm) [inline dbms/src/Functions/greatCircleDistance.cpp]: fmodf()?

**Comment by akuzm** [comment]
Let's merge the test first, so that we can see the performance improvement.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 4]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
     extern const int LOGICAL_ERROR;
 }
 
-static inline Float64 degToRad(Float64 angle) { return angle * DEGREES_IN_RADIANS; }
+static const double PI = 3.14159265358979323846;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: Authorship and license must be specified.

(Andrey Aksenov, the author of original code, permitted to use this code in ClickHouse under the Apache 2.0 license.)

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 5]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+static const float TO_RADF = static_cast<float>(PI / 180.0);
+static const float TO_RADF2 = static_cast<float>(PI / 360.0);
+
+const int GEODIST_TABLE_COS = 1024; // maxerr 0.00063%
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: static

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 6]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+const int GEODIST_TABLE_ASIN = 512;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: or anonymous namespace

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 7]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+const int GEODIST_TABLE_K = 1024;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: The name is not descriptive enough.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 8]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+
+static float g_GeoCos[GEODIST_TABLE_COS + 1];        ///< cos(x) table
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: Style.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 9]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+static float g_GeoAsin[GEODIST_TABLE_ASIN + 1];    ///< asin(sqrt(x)) table
+static float g_GeoFlatK[GEODIST_TABLE_K + 1][2];    ///< GeodistAdaptive() flat ellipsoid method k1,k2 coeffs table
+
+inline double sqr(double v)
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: static

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 10]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+{
+    return v*v;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: Style.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 11]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+}
+inline float fsqr(float v)
+{
+    return v*v;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: We can unify both methods.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 12]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+}
+
+void GeodistInit()
+{
+    for (int i = 0; i <= GEODIST_TABLE_COS; i++)
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: ++i

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 13]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+        g_GeoCos[i] = static_cast<float>(cos(2 * PI * i / GEODIST_TABLE_COS)); // [0, 2pi] -> [0, COSTABLE]
+
+    for (int i = 0; i <= GEODIST_TABLE_ASIN; i++)
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: int -> size_t

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 14]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+inline double sqr(double v)
+{
+    return v*v;
+}
+inline float fsqr(float v)
+{
+    return v*v;
+}
+
+void GeodistInit()
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: Style.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 15]
```diff
@@ -24,7 +21,91 @@ namespace ErrorCodes
...
+{
+    f = static_cast<float>(fabs(f));
+    while (f > 360)
+        f -= 360;
+    if (f > 180)
+        f = 360 - f;
+    return f;
+}
+
+static inline float GeodistFastCos(float x)
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: Style.

File: dbms/src/Functions/greatCircleDistance.cpp (hunk 4 above)
Comment (alexey-milovidov) [reply dbms/src/Functions/greatCircleDistance.cpp]: The link to presentation from HighLoad++ Siberia an a video is also valuable.

Comment (alexey-milovidov) [reply dbms/src/Functions/greatCircleDistance.cpp]: Also add a link to the original code.

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 16]
```diff
@@ -103,16 +184,30 @@ class FunctionGreatCircleDistance : public IFunction
...
+        float dlon = GeodistDegDiff(lon1Deg - lon2Deg);
+
+        if (dlon < 13)
+        {
+            // points are close enough; use flat ellipsoid model
+            // interpolate sqr(k1), sqr(k2) coefficients using latitudes midpoint
+            float m = (lat1Deg + lat2Deg + 180) * GEODIST_TABLE_K / 360; // [-90, 90] degrees -> [0, KTABLE] indexes
+            int i = static_cast<int>(m);
+            i &= (GEODIST_TABLE_K - 1);
+            float kk1 = g_GeoFlatK[i][0] + (g_GeoFlatK[i + 1][0] - g_GeoFlatK[i][0]) * (m - i);
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: If you have some explanation on algorithm, please provide it here (and rename variables accordingly).

File: dbms/src/Functions/greatCircleDistance.cpp [hunk 17]
```diff
@@ -103,16 +184,30 @@ class FunctionGreatCircleDistance : public IFunction
...
+            float kk2 = g_GeoFlatK[i][1] + (g_GeoFlatK[i + 1][1] - g_GeoFlatK[i][1]) * (m - i);
+            return static_cast<float>(sqrt(kk1 * dlat * dlat + kk2 * dlon * dlon));
+        }
+        // points too far away; use haversine
+        static const float D = 2 * 6371000;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/greatCircleDistance.cpp]: Constants should be also explained even if it's obvious for you.

File: dbms/tests/performance/great_circle_dist.xml [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+            <total_time_ms>10000</total_time_ms>
+        </any_of>
+    </stop_conditions>
+
+    <!-- lon [-180; 180], lat [-90; 90] -->
+    <query>SELECT count() FROM system.numbers WHERE NOT ignore(greatCircleDistance((rand() % 360) * 1. - 180, (number % 150) * 1.2 - 90, (number % 360) + toFloat64(rand()) / 4294967296 - 180, (rand() % 180) * 1. - 90))</query>
+    <!-- 55.755830, 37.617780 is center of Moscow -->
+    <query>SELECT count() FROM system.numbers WHERE NOT ignore(greatCircleDistance(55. + toFloat64(rand()) / 4294967296, 37. + toFloat64(rand()) / 4294967296, 55. + toFloat64(rand()) / 4294967296, 37. + toFloat64(rand()) / 4294967296))</query>
+</test>
+
```
Comment (alexey-milovidov) [inline dbms/tests/performance/great_circle_dist.xml]: And what is the improvement?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Comment by alexey-milovidov** [comment]
@stavrolia You did not answered what was the improvement.

**Comment by stavrolia** [comment]
@alexey-milovidov I have seen smth like 35x but maybe it was on my laptop, I don't remember

**Comment by alexey-milovidov** [comment]
Now it's about three times.

**Pull request merged**